File: apps/ios/Sources/License/LicenseCopy.swift
import Foundation

/// U+00B7 middle dot — one hidden character of the key.
private let licenseMaskCharacter: Character = "·"
/// The card shows the key's last group and masks everything before it.
private let licenseLastGroupLength = 4

/// What the License card renders, once per state.
///
/// Every field is a finished string for the view to place: the model decides
/// nothing about the license itself, and `status` is Rust's answer carried
/// through (the one exception is the Expired guard, which falls back to the
/// Unlicensed card rather than rendering a half-built date).
struct LicenseCardModel: Equatable {
    /// Already judged by Rust; never re-derived here.
    let status: LicenseStatus
    /// Unlicensed and Expired name themselves; Licensed is said by the coin.
    let badge: String?
    /// The purchase date, or `nil` when the activation carried none — the row
    /// is present and blank, and nothing is invented to fill it.
    let since: String?
    /// "Perpetual", "Valid until {date}" or "Expired {date}"; empty when there
    /// is no license to have a term.
    let term: String
    /// The stored key, masked to its last group. `nil` with no license.
    let maskedKey: String?
}

/// How a license state reads on the card. Pure: state in, strings out.
///
/// Presentation only — which catalog entry a state selects, which timestamps it
/// formats, and how the stored key is masked. Whether a license *is* licensed,
/// expired or invalid is decided in Rust (`futo-notes-license`, via
/// `licenseEvaluate`) and arrives already judged, so nothing here re-derives it
/// (AGENTS.md M6). Mirrors the desktop projection's
/// `src/features/license/licenseCopy.ts` and Android's `LicenseCopy.kt`.
func licenseCardModel(_ view: LicenseView, _ localization: Localization) -> LicenseCardModel {
    switch view.status {
    case .unlicensed:
        return unlicensedCard(localization)
    case .expired:
        // Only a v2 activation whose `expires_at` has passed reaches Expired,
        // so both dates are always there. The guard is kept, and kept identical
        // to the desktop and Android copy, so the three cannot drift over an
        // impossible case — and no card is ever dated "NaN".
        guard let issuedAt = view.issuedAtMillis, let expiresAt = view.expiresAtMillis else {
            return unlicensedCard(localization)
        }
        return LicenseCardModel(
            status: .expired,
            badge: localization.localizedText("license.statusExpired"),
            since: localization.localizedAbsoluteDate(Double(issuedAt)),
            term: localization.localizedText(
                "license.card.termExpired",
                arguments: ["date": localization.localizedAbsoluteDate(Double(expiresAt))]),
            maskedKey: maskedLicenseKey(view.key)
        )
    case .licensed:
        // A v1 activation carries no purchase time: `since` is nil, the row
        // renders blank, and no stand-in date is put in its place (decision
        // 2026-09-16 D2, issue #161). It is perpetual by format, not by guess.
        let term: String
        if let expiresAt = view.expiresAtMillis {
            term = localization.localizedText(
                "license.card.termValidUntil",
                arguments: ["date": localization.localizedAbsoluteDate(Double(expiresAt))])
        } else {
            term = localization.localizedText("license.card.termPerpetual")
        }
        return LicenseCardModel(
            status: .licensed,
            // Licensed wears no badge: the coin in the well is the statement.
            badge: nil,
            since: view.issuedAtMillis.map { localization.localizedAbsoluteDate(Double($0)) },
            term: term,
            maskedKey: maskedLicenseKey(view.key)
        )
    }
}

private func unlicensedCard(_ localization: Localization) -> LicenseCardModel {
    LicenseCardModel(
        status: .unlicensed,
        badge: localization.localizedText("license.unlicensed"),
        since: nil,
        term: "",
        maskedKey: nil
    )
}

/// The stored key with every character before its last group replaced by a
/// middle dot, and its hyphens left where they are:
/// `····-····-····-····-····-····-····-6UJV`.
///
/// The mask is the SAME LENGTH as the key and keeps its separators in the same
/// columns, which is what makes revealing it an in-place swap: the card renders
/// the value in a monospace face, so every dot is replaced by the character that
/// was hiding under it and nothing on the plate moves (@justin 2026-09-18). The
/// first mask was a fixed seven groups of dots joined by SPACES — 39 characters
/// whatever the key, so an org-prefixed 42-character key jumped three cells to
/// the right as it appeared, and the rows below it shifted with it. The key
/// arrives normalized (trimmed, uppercased) from the Rust crate, so it is masked
/// as it stands.
private func maskedLicenseKey(_ key: String?) -> String? {
    guard let key else { return nil }
    guard key.count > licenseLastGroupLength else { return key }
    let hidden = key.dropLast(licenseLastGroupLength).map {
        $0 == "-" ? $0 : licenseMaskCharacter
    }
    return String(hidden) + String(key.suffix(licenseLastGroupLength))
}

File: src/features/license/licenseCopy.ts
// How a license state reads on the card. Pure: state in, strings out.
//
// This is presentation only — which catalog entry a state selects, which
// timestamps it formats, and how the stored key is masked. Whether a license
// *is* licensed, expired or invalid is decided in Rust (`futo-notes-license`)
// and arrives already judged, so nothing here re-derives it (AGENTS.md M6).

import type { LicenseStateName, LicenseView } from '$lib/platform/license';
import { localizedAbsoluteDate, localizedText } from '$shared/localization';

/** U+00B7 middle dot — one hidden character of the key. */
const MASK_CHARACTER = '·';
/** The card shows the key's last group and masks everything before it. */
const LAST_GROUP_LENGTH = 4;

function date(iso: string): string {
  return localizedAbsoluteDate(Date.parse(iso));
}

/// The stored key with every character before its last group replaced by a
/// middle dot, and its hyphens left where they are:
/// `····-····-····-····-····-····-····-6UJV`.
///
/// The mask is the SAME LENGTH as the key and keeps its separators in the same
/// columns, which is what makes revealing it an in-place swap: the card renders
/// the value in a monospace face, so every dot is replaced by the character
/// that was hiding under it and nothing on the plate moves (@justin
/// 2026-09-18). The first mask was a fixed seven groups of dots joined by
/// SPACES — 39 characters whatever the key, so an org-prefixed 42-character key
/// jumped three cells to the right as it appeared, and the rows below it
/// shifted with it. The key arrives normalized (trimmed, uppercased) from the
/// Rust crate, so it is masked as it stands.
function maskKey(key: string): string {
  if (key.length <= LAST_GROUP_LENGTH) return key;
  const hidden = key.slice(0, -LAST_GROUP_LENGTH).replace(/[^-]/g, MASK_CHARACTER);
  return hidden + key.slice(-LAST_GROUP_LENGTH);
}

/// What the License card renders, once per state.
///
/// Every field is a finished string for the shell to place: the card decides
/// nothing about the license itself, and `status` is Rust's answer carried
/// through (the one exception is the Expired guard below, which falls back to
/// the Unlicensed card rather than rendering a half-built date).
export interface LicenseCardModel {
  status: LicenseStateName;
  /** Unlicensed and Expired name themselves; Licensed is said by the coin. */
  badge: string | null;
  /** The purchase date, or `null` when the activation carried none — the row
   *  is present and blank, and nothing is invented to fill it. */
  since: string | null;
  /** "Perpetual", "Valid until {date}" or "Expired {date}"; empty when there
   *  is no license to have a term. */
  term: string;
  /** The stored key, masked to its last group. `null` with no license. */
  maskedKey: string | null;
}

function unlicensedCard(): LicenseCardModel {
  return {
    status: 'unlicensed',
    badge: localizedText('license.unlicensed'),
    since: null,
    term: '',
    maskedKey: null,
  };
}

/// The ambient label — the thing a purchase removes. "Unlicensed" while
/// Unlicensed *or* Expired; "Licensed since {date}" once Licensed.
///
/// `null` means **render nothing at all**: a v1 activation carries no
/// `issued_at`, so there is no date for "Licensed since" and no dateless
/// variant of that line (decision 2026-09-10, reaffirmed as D2). The footer is
/// then empty — dropping the "Unlicensed" label is the whole visible reward,
/// and it still happens.
export function licenseAmbientLabel(view: LicenseView): string | null {
  if (view.state === 'licensed') {
    return view.issuedAt === null
      ? null
      : localizedText('license.licensedSince', { date: date(view.issuedAt) });
  }
  return localizedText('license.unlicensed');
}

/// The License card's fields.
///
/// A clause is dropped when the activation did not carry its field, never
/// filled in with a stand-in: a v1 activation has no `issued_at`, so `since` is
/// `null` and the row renders blank, and no `expires_at`, so the term is
/// "Perpetual" — which is what a v1 license is by format, not a guess.
export function licenseCardModel(view: LicenseView): LicenseCardModel {
  if (view.state === 'unlicensed') return unlicensedCard();

  if (view.state === 'expired') {
    // Expired is only ever reached by a v2 activation whose `expires_at` has
    // passed, so it always has both dates. The guard stays — and stays
    // identical on all three platforms — so the impossible case cannot render
    // a card dated "NaN" if the contract ever changes.
    if (view.issuedAt === null || view.expiresAt === null) return unlicensedCard();
    return {
      status: 'expired',
      badge: localizedText('license.statusExpired'),
      since: date(view.issuedAt),
      term: localizedText('license.card.termExpired', { date: date(view.expiresAt) }),
      maskedKey: maskedKeyOf(view),
    };
  }

  return {
    status: 'licensed',
    // Licensed wears no badge: the coin in the well is the statement.
    badge: null,
    since: view.issuedAt === null ? null : date(view.issuedAt),
    // A perpetual product has no expiry, and the term must not invent one. A
    // v1 activation lands here too: it is perpetual by format, not by guess.
    term:
      view.expiresAt === null
        ? localizedText('license.card.termPerpetual')
        : localizedText('license.card.termValidUntil', { date: date(view.expiresAt) }),
    maskedKey: maskedKeyOf(view),
  };
}

function maskedKeyOf(view: LicenseView): string | null {
  return view.key === null ? null : maskKey(view.key);
}
